README.md
# 《标题日记》 后台 - nodejs 版


## 一、项目说明
该后台服务的项目：  
- [《标题日记》](https://github.com/KyleBing/diary-vue)
- [《路书》](https://github.com/KyleBing/map)
- 《我的后台》`未开源`
- 《挪车二维码》`未开源`

> 线上已运行的例子：  
> http [http://kylebing.cn:3000/diary/detail?diaryId=6766](http://kylebing.cn:3000/diary/detail?diaryId=6766)



### 
```bash
# 1. 统计信息
- /statistic # 统计信息 `2022-05-12`

# 2. 日记管理
- /diary/list
- /diary/add
- /diary/modify
- /diary/delete
- /diary/detail

    # 2.1 银行卡列表信息
    - /bank-card/ # 银行卡列表 `2022-05-12`
    
    # 2.2 账单统计信息
    - /bill/ # 日记中的账单类别统计信息 `2022-05-24`
    - /bill/keys # 所有账单条目列表 `2023-12-20`
    - /bill/sorted # 所有账单展示

# 3. 用户管理
- /user/
- /user/add
- /user/delete
- /user/modify
- /user/detail
- /user/list
- /user/login
- /user/change-password

# 4. 日记类别管理
- /diary-category/ # 类别管理 `2022-05-24`
- /diary-category/add # 类别管理 `2022-05-24`
- /diary-category/modify # 类别修改 `2022-07-22`
- /diary-category/list # 类别管理 `2022-05-24`
- /diary-category/delete

# 5. 邀请码管理
- /invitation/generate # 生成新的
- /invitation/mark-shared # 标记邀请码为已用
- /invitation/list # 可用的邀请码列表
- /invitation/delete # 删除邀请码

# 6. 统计
- /statistic/ # 统计日记、用户数据
- /statistic/category # 类别日记数量统计
- /statistic/year # 年份日记统计
- /statistic/users # 用户的日记、码表、qr 等数量
- /statistic/weather # 所有日记的气温信息

# 7. 路书数据
地图路线
- /map-route/list
- /map-route/detail
- /map-route/add
- /map-route/modify
- /map-route/delete

地图信息
- /map-pointer/list
- /map-pointer/detail
- /map-pointer/add
- /map-pointer/modify
- /map-pointer/delete

```



## 二、安装说明

**服务器需要的条件：**
- 已安装 `nodejs 18+`
- 已安装 `npm` 或 `yarn`
- 已安装 `nginx`，需要用它进行路径映射，以供前端非跨域式访问后台
- 已安装 `mysql` 或 `mariaDB`

### 1. clone 或 下载该项目文件
下载到项目文件后，执行 `npm i` 或者 `yarn` 安装项目依赖

### 2. 修改数据库配置文件
修改 `/config/configDatabase.js` 文件内容，改成你的配置
```js
module.exports = {
    host:       'localhost',
    user:       'root',
    password:   '----',
    port:       '3306',
    database:   'diary',
    multipleStatements: true
}
```

和项目配置文件 `/config/configProject.js`
```js
module.exports = {
   invitation: '----', // 万能邀请码，用于注册时使用
   adminCount: 'kylebing@163.com', // 管理员帐户，该用户可以在统计页面中查看所有用户统计数据
   TOKEN_NAME: 'Diary-Token', // header 中的 token 字段，跟前端约定的 token 名字，不需要刻意修改
}
```

### 3. 启动程序
这里推荐使用 pm2 管理程序，创建名为 `diary` 的 pm2 项目，并启动
> pm2 的使用方法： [https://blog.csdn.net/KimBing/article/details/124249590](https://blog.csdn.net/KimBing/article/details/124249590)

```bash
pm2 start bin/www --name diary
```

如果你实在不用 pm2，也可以直接使用 `npm` 原始方法启动

```bash
npm run start
```

项目启动后会运行在 `localhost:3000`，直接访问这个地址应该能看到：

<img width="539" alt="Screen Shot 2022-04-19 at 21 47 25" src="https://user-images.githubusercontent.com/12215982/164018379-bb497ec3-53e4-46c5-969a-c5b3ca4c0c31.png">

###docker方式启动：
1. 首先确保系统安装了docker，没有的话执行：
   ```sh
   bash <(curl -sSL https://get.docker.com)
   ```
2. Start the Service

   ```sh
   docker compose up -d   

   或者：

   ```sh
   docker run -itd \
    -v $PWD/config/:/app/config/ \
    --restart=unless-stopped \
    --name diary-portal \
    ghcr.io/gallexy/diary-portal:latest
   ```    


### 4. 初始化数据库

> 注意：初始化会清空 `diary` 数据库中的所有内容


1. 删除后台目录中的 `DATABASE_LOCK` 这个文件。  
2. 直接访问 `你服务器的域名或IP:3000/init` 这个路径即可将数据库初始化，初始化数据库会自动创建一个为名 `diary` 的数据库。
3. 初始化后，会自动在项目目录中新建一个名为 `DATABASE_LOCK` 的文件，之后将不能再执行这个接口，如果想再次初始化，需要先删除这个文件。



### 5. 配置 nginx，映射 `localhost:3000` 路径到  `/portal` 路径

1. 打开 nginx 的配置文件，
- CentOS 的 nginx 配置文件在 `/etc/nginx/conf.d/` 目录下。
- Ubuntu 的 nginx 配置文件在 `/etc/nginx/site-avilable/default` 中。

2. 打开 `default.conf` 或 `default` 文件
    ```bash
    vi default.conf
    ```

3. 在 http 内部， server 外部，添加以下内容

    ```bash
     upstream diary_server {
         server localhost:3000;
         keepalive 2000;
     }
    ```

4. 然后在 server 内部添加：

    ```bash
    location /portal/ {
        proxy_pass http://diary_server/; # 这里对应上面的 upstream 名字
        proxy_set_header Host $host:$server_port; # 这里照搬就可以
    }
    ```
5. 这样，就会将 `localhost:3000` 这个接口映射到 `localhost/portal/` 这个路径下
6. 重启 nginx 服务
    ```bash
    systemctl restart nginx
    ```
7. 添加 cron 定时任务
   用户数据里有对用户的日记和其它信息的统计，这个统计过程耗时稍长，所以将其设成定时任务，每小时执行一次。
   以 Ubuntu 为例
   执行
   ```bash
   crontab -e
   ```
   然后添加以下内容到打开的窗口中，意思就是说第小时的 17 分统计并更新用户数据，下面的 js 路径改成自己系统中的 JS 路径。
   ```bash
   17 * * * * node /var/www/html/portal/routes/cron/updateUserInfo.js
   ```
   然后重启 cron 服务
   ```bash
   systemctl restart cron
   ```

### 6. 设置管理员账户
1. 用上面设置的通用邀请码注册之后
2. 从数据库中直接修改对应用户的 `users.group_id`  改为 `1` (管理员)
3. 网页上重新登录该用户，就能看到 **邀请码** 菜单了。

### 7. 配置前端项目
1. 下载 [https://github.com/KyleBing/diary-vue](https://github.com/KyleBing/diary-vue)
2. 安装依赖，执行 `npm i` 或者 `yarn`
3. 如果你需要修改前端请求后台的路径，修改 `/src/request.js` 中的 `BASE_URL` 即可
4. 执行 `npm build` 会生成前端项目的生产环境的文件，就这个项目而言，它会在生成在 `../diary/` 目录，也会在 `./archive` 目录下生成一个名为 `diary-2023-06-xx.zip` 的压缩包，这个压缩包的内容就是 `../diary` 的内容，但不包含外层 `diary` 目录。
5. 将项目文件放置于服务器 nginx 主目录的 `/diary/` 下
6. 此时后台项目在 `/portal/` 目录下，前端项目在 `/dairy/` 下，这样就能直接使用了

## 三、开发说明

### 1. 密码说明
密码使用 [bcrypt](https://github.com/kelektiv/node.bcrypt.js) 加密
### 2. 返回数据格式

```json
{
  "success": true,
  "message": "提示信息",
  "data": {}
}
```

## 四、其它
> 始于 `2022-04-14`


CHANGELOG.md
# CHANGELOG



### v1.00 `2022-06-22`
- 添加账单数据，月份分类


### 创建 docker-ready 分支
-把utility.js从config目录剥离到lib目录，config目录只保留需要的配置信息，启动docker时用户可以提供自己的配置

-mysql更新到mysql2,避免认证错误

-增加Dockerfile和docker-compose.yml，用户可以以docker方式启动服务，甚至可以制作自己的docker镜像


routes/user/user.js
const express = require('express')
const configProject = require('../../config/configProject')
const utility = require("../../lib/utility");
const ResponseSuccess = require("../../response/ResponseSuccess");
const ResponseError = require("../../response/ResponseError");
const router = express.Router()
const bcrypt = require('bcrypt')




/* GET users listing. */
router.post('/register', (req, res, next) => {
    // TODO: 验证传过来的数据库必填项
    if (req.body.invitationCode === configProject.invitation){ // 万能全局邀请码
        registerUser(req, res)
    } else {
        utility
            .getDataFromDB('diary', [`select * from invitations where id = '${req.body.invitationCode}'`], true)
            .then(result => {
                if (result){
                    if (result.binding_uid){
                        res.send(new ResponseError('', '邀请码已被使用'))
                    } else {
                        registerUser(req, res)
                    }
                } else {
                    res.send(new ResponseError('', '邀请码无效'))
                }
            })
            .catch(err => {
                res.send(new ResponseError(err, '数据库请求出错'))
            })
    }
})

function registerUser(req, res){
    checkEmailOrUserNameExist(req.body.email, req.body.username)
        .then(dataEmailExistArray => {
            // email 记录是否已经存在
            if (dataEmailExistArray.length > 0){
                return res.send(new ResponseError('', '邮箱或用户名已被注册'))
            } else {
                let sqlArray = []
                let timeNow = utility.dateFormatter(new Date())
                // 明文密码通过 bcrypt 加密，对比密码也是通过  bcrypt
                bcrypt.hash(req.body.password, 10, (err, encryptPassword) => {
                    sqlArray.push(
                        // 注册的用户默认为普通用户
                        `insert into users(email, nickname, username, password, register_time, last_visit_time, comment, 
                                                wx, phone, homepage, gaode, group_id)
                                    VALUES (
                                    '${req.body.email}', 
                                    '${req.body.nickname || ''}', 
                                    '${req.body.username || ''}', 
                                    '${encryptPassword}', 
                                    '${timeNow}',
                                    '${timeNow}',
                                    '${req.body.comment || ''}', 
                                    '${req.body.wx || ''}', 
                                    '${req.body.phone || ''}', 
                                    '${req.body.homepage || ''}', 
                                    '${req.body.gaode || ''}', 
                                    '2'
                                    )`
                    )
                    utility
                        .getDataFromDB( 'diary', sqlArray)
                        .then(data => {
                            let lastInsertedUid = data.insertId
                            utility.getDataFromDB('diary', [`update invitations set binding_uid = ${lastInsertedUid}, date_register = '${timeNow}' where id = '${req.body.invitationCode}'`])
                                .then(resInvitation => {
                                    res.send(new ResponseSuccess('', '注册成功'))
                                })
                                .catch(err => {
                                    res.send(new ResponseError('', '注册成功，邀请码信息更新失败'))
                                })
                        })
                        .catch(err => {
                            res.send(new ResponseError('', '注册失败'))
                        })
                })

            }
        })
        .catch(errEmailExist => {
            console.log(errEmailExist)
            res.send(new ResponseError(errEmailExist, '查询出错'))
        })
}

// 检查用户名或邮箱是否存在
function checkEmailOrUserNameExist(email, username){
    let sqlArray = []
    sqlArray.push(`select * from users where email='${email}' or username ='${username}'`)
    return utility.getDataFromDB( 'diary', sqlArray)
}

router.post('/list', (req, res, next) => {
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            let promisesAll = []

            if (userInfo.group_id === 1){
                // admin user
                let pointStart = (Number(req.body.pageNo) - 1) * Number(req.body.pageSize)
                promisesAll.push(utility.getDataFromDB(
                    'diary',
                    [`SELECT * from users limit ${pointStart} , ${req.body.pageSize}`])
                )
                promisesAll.push(utility.getDataFromDB(
                    'diary',
                    [`select count(*) as sum from users`], true)
                )
            } else {
                // normal user
                let pointStart = (Number(req.body.pageNo) - 1) * Number(req.body.pageSize)
                promisesAll.push(utility.getDataFromDB(
                    'diary',
                    [`SELECT * from users where uid = '${userInfo.uid}' limit ${pointStart} , ${req.body.pageSize}`])
                )
                promisesAll.push(utility.getDataFromDB(
                    'diary',
                    [`select count(*) as sum from users where uid = '${userInfo.uid}' `], true)
                )
            }


            Promise
                .all(promisesAll)
                .then(([userList, dataSum]) => {
                    utility.updateUserLastLoginTime(userInfo.uid)
                    userList.forEach(diary => {
                        // decode unicode
                        diary.title = utility.unicodeDecode(diary.title)
                        diary.content = utility.unicodeDecode(diary.content)
                    })
                    res.send(new ResponseSuccess({
                        list: userList,
                        pager: {
                            pageSize: Number(req.body.pageSize),
                            pageNo: Number(req.body.pageNo),
                            total: dataSum.sum
                        }
                    }, '请求成功'))

                })
                .catch(err => {
                    res.send(new ResponseError(err, err.message))
                })
        })
        .catch(errInfo => {
            res.send(new ResponseError('', errInfo))
        })
})

router.get('/detail', (req, res, next) => {
    let sqlArray = []
    sqlArray.push(`select * from qrs where hash = '${req.query.hash}'`)
    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            // decode unicode
            data.message = utility.unicodeDecode(data.message)
            data.description = utility.unicodeDecode(data.description)

            // 2. 判断是否为共享 QR
            if (data.is_public === 1){
                // 2.1 如果是，直接返回结果，不需要判断任何东西
                res.send(new ResponseSuccess(data))
            } else {
                // 2.2 如果不是，需要判断：当前 email 和 token 是否吻合
                utility
                    .verifyAuthorization(req)
                    .then(userInfo => {
                        // 3. 判断 QR 是否属于当前请求用户
                        if (Number(userInfo.uid) === data.uid){
                            // 记录最后访问时间
                            utility.updateUserLastLoginTime(userInfo.uid)
                            /*                            // TODO:过滤可见信息 自己看，管理员看，其它用户看
                                                        if (data.is_show_wx){
                                                            data.wx = ''
                                                        }
                                                        if (data.is_show_car){
                                                            data.car = ''
                                                            data.car_desc = ''
                                                            data.car_plate = ''
                                                        }
                                                        if (data.is_show_gaode){
                                                            data.gaode = ''
                                                        }
                                                        if (data.is_show_homepage){
                                                            data.homepage = ''
                                                        }*/
                            res.send(new ResponseSuccess(data))
                        } else {
                            res.send(new ResponseError('','当前用户无权查看该 QR ：请求用户 ID 与 QR 归属不匹配'))
                        }
                    })
                    .catch(errInfo => {
                        res.send(new ResponseError('', errInfo))
                    })
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, err.message))
        })
})


router.get('/avatar', (req, res, next) => {
    let sqlArray = []
    sqlArray.push(`select avatar from users where email = '${req.query.email}'`)
    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            res.send(new ResponseSuccess(data))
        })
        .catch(err => {
            res.send(new ResponseError(err, err.message))
        })
})


router.post('/add', (req, res, next) => {
    checkEmailOrUserNameExist(req.body.email, req.body.username)
        .then(dataEmailExistArray => {
            // email 记录是否已经存在
            if (dataEmailExistArray.length > 0){
                return res.send(new ResponseError('', '邮箱或用户名已被注册'))
            } else {
                let sqlArray = []
                let timeNow = utility.dateFormatter(new Date())
                // 明文密码通过 bcrypt 加密，对比密码也是通过  bcrypt
                bcrypt.hash(req.body.password, 10, (err, encryptPassword) => {
                    sqlArray.push(
                        `insert into users(email, nickname, username, password, register_time, last_visit_time, comment, 
                                                wx, phone, homepage, gaode, group_id)
                                    VALUES (
                                    '${req.body.email}', 
                                    '${req.body.nickname}', 
                                    '${req.body.username}', 
                                    '${encryptPassword}', 
                                    '${timeNow}',
                                    '${timeNow}',
                                    '${req.body.comment || ''}', 
                                    '${req.body.wx}', 
                                    '${req.body.phone}', 
                                    '${req.body.homepage}', 
                                    '${req.body.gaode}', 
                                    '${req.body.group_id}'
                                    )`
                    )
                    utility
                        .getDataFromDB( 'diary', sqlArray)
                        .then(data => {
                            res.send(new ResponseSuccess('', '用户添加成功'))
                        })
                        .catch(err => {
                            res.send(new ResponseError(err, '用户添加失败'))
                        })
                })

            }
        })
        .catch(errEmailExist => {
            console.log(errEmailExist)
            res.send(new ResponseError(errEmailExist, '查询出错'))
        })
})

// 检查用户名或邮箱是否存在
function checkHashExist(username, email){
    let sqlArray = []
    sqlArray.push(`select * from users where username ='${username.toLowerCase()}' or email = '${email}'`)
    return utility.getDataFromDB( 'diary', sqlArray)
}


// 设置用户资料：昵称，avatar，手机号
router.put('/set-profile', (req, res, next) => {
    // 1. 验证用户信息是否正确
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            let avatar = req.body.avatar || ''
            let sqlArray = []
            sqlArray.push(`
                update users
                set users.nickname       = '${req.body.nickname}',
                    users.phone          = '${req.body.phone}',
                    users.avatar         = '${avatar}',
                    users.city           = '${req.body.city}',
                    users.geolocation    = '${req.body.geolocation}'
                    WHERE uid = '${userInfo.uid}'
            `)
            utility
                .getDataFromDB('diary', sqlArray, true)
                .then(data => {
                    utility.updateUserLastLoginTime(userInfo.uid)
                    utility
                        .verifyAuthorization(req)
                        .then(newUserInfo => {
                            res.send(new ResponseSuccess(newUserInfo, '修改成功'))
                        })
                })
                .catch(err => {
                    res.send(new ResponseError(err, '修改失败'))
                })
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })
})


router.put('/modify', (req, res, next) => {

    // 1. 验证用户信息是否正确
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.group_id === 1) {
                operateUserInfo(req, res, userInfo)
            } else {
                if (userInfo.uid !== req.body.uid){
                    res.send(new ResponseError('', '你无权操作该用户信息'))
                } else {
                    operateUserInfo(req, res, userInfo)
                }
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })
})

function operateUserInfo(req, res, userInfo){
    let sqlArray = []
    sqlArray.push(`
                update users
                    set
                            users.email = '${req.body.email}', 
                            users.nickname = '${req.body.nickname}', 
                            users.username = '${req.body.username}', 
                            users.comment = '${req.body.comment || ''}', 
                            users.wx = '${req.body.wx}', 
                            users.phone = '${req.body.phone}', 
                            users.homepage = '${req.body.homepage}', 
                            users.gaode = '${req.body.gaode}', 
                            users.group_id = '${req.body.group_id}'
                    WHERE uid='${req.body.uid}'
            `)

    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            utility.updateUserLastLoginTime(userInfo.uid)
            res.send(new ResponseSuccess(data, '修改成功'))
        })
        .catch(err => {
            res.send(new ResponseError(err, '修改失败'))
        })
}

router.delete('/delete', (req, res, next) => {
    // 1. 验证用户信息是否正确
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.group_id === 1){
                let sqlArray = []
                sqlArray.push(`
                        DELETE from users
                        WHERE uid='${req.body.uid}'
                    `)
                utility
                    .getDataFromDB( 'diary', sqlArray)
                    .then(data => {
                        if (data.affectedRows > 0) {
                            utility.updateUserLastLoginTime(req.body.email)
                            res.send(new ResponseSuccess('', '删除成功'))
                        } else {
                            res.send(new ResponseError('', '删除失败'))
                        }
                    })
                    .catch(err => {
                        res.send(new ResponseError(err,))
                    })
            } else {
                res.send(new ResponseError('', '无权操作'))
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })
})

router.post('/login', (req, res, next) => {
    let sqlArray = []
    sqlArray.push(`select * from users where email = '${req.body.email}'`)

    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            if (data) {
                bcrypt.compare(req.body.password, data.password, function(err, isPasswordMatch) {
                    if (isPasswordMatch){
                        utility.updateUserLastLoginTime(req.body.email)
                        res.send(new ResponseSuccess(data,'登录成功'))
                    } else {
                        res.send(new ResponseError('','用户名或密码错误'))
                    }
                })
            } else {
                res.send(new ResponseError('', '无此用户'))
            }

        })
        .catch(err => {
            res.send(new ResponseError('', err.message))
        })
})

// 修改密码
router.put('/change-password', (req, res, next) => {
    if (!req.body.password){
        res.send(new ResponseError('', '参数错误：password 未定义'))
        return
    }

    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.email === 'test@163.com'){
                res.send(new ResponseError('', '演示帐户密码不允许修改'))
                return
            }
            bcrypt.hash(req.body.password, 10, (err, encryptPasswordNew) => {
                let changePasswordSqlArray = [`update users set password = '${encryptPasswordNew}' where email='${userInfo.email}'`]
                utility
                    .getDataFromDB( 'diary', changePasswordSqlArray)
                    .then(dataChangePassword => {
                        utility.updateUserLastLoginTime(userInfo.uid)
                        res.send(new ResponseSuccess('', '修改密码成功'))
                    })
                    .catch(errChangePassword => {
                        res.send(new ResponseError('', '修改密码失败'))
                    })
            })
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })

})

// 注销帐号
router.delete('/destroy-account', (req, res, next) => {
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            // 演示帐户时不允许执行注销操作
            if (userInfo.email === 'test@163.com'){
                res.send(new ResponseError('', '演示帐户不允许执行此操作'))
                return
            }
            let connection = utility.getMysqlConnection('diary')
            connection.beginTransaction(transactionError => {
                if (transactionError){
                    connection.rollback(err => {
                        res.send(new ResponseError('', 'beginTransaction: 事务执行失败，已回滚'))
                    })
                    connection.end()
                } else {
                    let sql = `
                                delete from diaries where uid = ${userInfo.uid}; 
                                delete from invitations where binding_uid = ${userInfo.uid}; 
                                delete from map_pointer where uid = ${userInfo.uid}; 
                                delete from map_route where uid = ${userInfo.uid}; 
                                delete from map_route where uid = ${userInfo.uid}; 
                                delete from qrs where uid = ${userInfo.uid}; 
                                delete from users where uid = ${userInfo.uid}; 
                                `
                    connection.query(sql, [], (queryErr,result) => {
                        if (queryErr){
                            connection.rollback(err => {
                                res.send(new ResponseError(queryErr, 'query: 事务执行失败，已回滚'))
                            })
                            // res.send(new ResponseError(err, '数据库请求错误'))
                        } else {
                            connection.commit(commitError => {
                                if (commitError){
                                    connection.rollback(err => {
                                        res.send(new ResponseError(err, 'transaction.commit: 事务执行失败，已回滚'))
                                    })
                                } else {
                                    res.send(new ResponseSuccess(result, '事务执行成功'))
                                }
                            })
                        }
                        connection.end()
                    })
                }

            })
        })
        .catch(errInfo => {
            res.send(new ResponseError('null', errInfo))
        })
})


module.exports = router
